### PR TITLE
Move nomad query parsing

### PR DESCRIPTION
**Pull request by lgfa29**
Moving the query parsing logic to the Nomad APM plugin keeps it closer to where it actually is used.

### CODE REVIEW

File: plugins/builtin/apm/nomad/plugin/metrics.go [hunk 1]
```diff
@@ -91,3 +95,47 @@ func (a *APMPlugin) Query(q string) (float64, error) {
...
+		Group: mainParts[2],
+	}
+
+	opMetricParts := strings.Split(mainParts[0], "_")
+	if len(opMetricParts) < 2 {
+		return nil, fmt.Errorf(`expected <operation>_<metric>, received "%s"`, mainParts[0])
+	}
+
+	op := opMetricParts[0]
+	metric := strings.Join(opMetricParts[1:], "_")
```
[the remark this answers is not in the document]
Comment (lgfa29) [reply plugins/builtin/apm/nomad/plugin/metrics.go]: TIL. Very useful function.

File: plugins/builtin/apm/nomad/plugin/metrics.go [hunk 2]
```diff
@@ -91,3 +95,47 @@ func (a *APMPlugin) Query(q string) (float64, error) {
 
 	return result, nil
 }
+
+func parseQuery(q string) (*Query, error) {
+	mainParts := strings.Split(q, "/")
+	if len(mainParts) != 3 {
```
[the remark this answers is not in the document]
Comment (lgfa29) [reply plugins/builtin/apm/nomad/plugin/metrics.go:101]: Since only jobs can have forward slashes I just switched the order of group and job and used `SplitN`. It's an internal representational, so changing it shouldn't be an issue.

File: plugins/builtin/apm/nomad/plugin/metrics.go [hunk 3]
```diff
@@ -91,3 +95,47 @@ func (a *APMPlugin) Query(q string) (float64, error) {
...
+
+	switch op {
+	case "sum":
+		fallthrough
+	case "avg":
+		fallthrough
+	case "min":
+		fallthrough
+	case "max":
+		query.Operation = op
```
Comment (jrasell) [inline plugins/builtin/apm/nomad/plugin/metrics.go:135]: for me this is a bit easier to read and more concise

```suggestion
	case "sum", "avg", "min", "max":
		query.Operation = op
```

File: policy/nomad/parser.go [hunk 4]
```diff
@@ -103,19 +102,9 @@ func canonicalizePolicy(from *api.ScalingPolicy, to *policy.Policy) {
...
-		switch metric {
-		case "cpu":
-			metric = "nomad.client.allocs.cpu.total_percent"
-		case "memory":
-			metric = "nomad.client.allocs.memory.usage"
-		}
-
-		to.Query = fmt.Sprintf("%s/%s/%s/%s", metric, from.Target["Job"], from.Target["Group"], op)
+	if to.Source == plugins.InternalAPMNomad {
+		to.Query = fmt.Sprintf("%s/%s/%s", to.Query, from.Target["Group"], from.Target["Job"])
```
Comment (jrasell) [inline policy/nomad/parser.go:108]: curious if this needs to be its own conditional or could it be moved above?

Comment (lgfa29) [reply policy/nomad/parser.go:108]: This needs to be a separate conditional for the case when `to.Source` is defined by the operator as already being `plugins.InternalAPMNomad`, for example:

```hcl
...
scaling {
  ...
  policy {
    source = "nomad-apm"
    query  = "avg_cpu"
    ...
  }
}
```

It's a bit of an edge case, but we should cover it regardless. I am writing tests for the parser in a separate PR, so I will make sure to create one to cover this.

Comment (jrasell) [reply policy/nomad/parser.go:108]: ohhh yeah!
